src/graph/graph_op.cc

```diff
@@ -653,4 +654,77 @@ DGL_REGISTER_GLOBAL("graph_index._CAPI_DGLMapSubgraphNID")
     *rv = GraphOp::MapParentIdToSubgraphId(parent_vids, query);
   });
 
+#if !defined(WIN32) && !defined(_WIN32) && !defined(__WIN32__) && !defined(__NT__)
+
+IdArray GraphOp::MetisPartition(GraphPtr g, int k) {
+  // The index type of Metis needs to be compatible with DGL index type.
+  assert(sizeof(idx_t) == sizeof(dgl_id_t));
+  ImmutableGraphPtr ig = std::dynamic_pointer_cast<ImmutableGraph>(g);
+  if (ig == NULL) {
+    std::cerr << "The input graph isn't immutable graph" << std::endl;
+    return aten::NewIdArray(0);
+  }
+  // This is a symmetric graph, so in-csr and out-csr are the same.
+  CSRPtr csr = ig->GetInCSR();
+  auto mat = csr->ToCSRMatrix();
+
+  idx_t nvtxs = g->NumVertices();
+  idx_t ncon = 1;        // # balacing constraints.
+  idx_t *xadj = static_cast<idx_t*>(mat.indptr->data);
+  idx_t *adjncy = static_cast<idx_t*>(mat.indices->data);
+  // TODO(zhengda) we could use this to create partitions that balance both #vertices and #edges.
+  idx_t *vwgt = NULL;
+  idx_t nparts = k;
+  IdArray part_arr = aten::NewIdArray(nvtxs);
+  idx_t objval = 0;
+  idx_t *part = static_cast<idx_t*>(part_arr->data);
+  int ret = METIS_PartGraphKway(&nvtxs,      // The number of vertices
+                                &ncon,       // The number of balancing constraints.
+                                xadj,        // indptr
+                                adjncy,      // indices
+                                NULL,        // the weights of the vertices
+                                NULL,        // The size of the vertices for computing
+                                // the total communication volume
+                                NULL,        // The weights of the edges
+                                &nparts,     // The number of partitions.
+                                NULL,        // the desired weight for each partition and constraint
+                                NULL,        // the allowed load imbalance tolerance
+                                NULL,        // the array of options
+                                &objval,      // the edge-cut or the total communication volume of
+                                // the partitioning solution
+                                part);
```

Comment:
I wonder does METIS use a zero-based CSR or a one-based?

Reply:
zero-based mean the first element in indptr is 0?